File: backend/services/advanced_booking_service.py
# backend/services/advanced_booking_service.py - Advanced booking business logic (MongoDB)

from datetime import datetime
from typing import List, Optional, Dict, Any
import uuid
from database import mongodb
from services import user_service

def _now_iso() -> str:
    return datetime.utcnow().isoformat() + "Z"

def _gen_id(prefix: str = "") -> str:
    return f"{prefix}{uuid.uuid4().hex[:12].upper()}"

def _doctor_matches_case(doctor: dict, case_type: str) -> bool:
    category = str(doctor.get('advanced_booking_category') or doctor.get('advancedBookingCategory') or 'general').lower()
    case_key = str(case_type or '').lower()
    if category == case_key:
        return True
    if category not in ('general', 'any', ''):
        return False

    text = f"{doctor.get('specialization', '')} {doctor.get('department', '')}".lower()
    if case_key == 'pregnancy':
        return any(token in text for token in ('gyn', 'obst', 'women', 'maternity'))
    if case_key == 'baby':
        return any(token in text for token in ('pedia', 'child', 'neonatal'))
    if case_key == 'elder':
        return any(token in text for token in ('geriat', 'general', 'internal', 'medicine', 'senior'))
    return True

def _doctor_is_available(doctor: dict) -> bool:
    status = str(doctor.get('status') or 'active').lower()
    return status not in ('inactive', 'disabled', 'blocked', 'suspended')

async def _allocate_doctor(hospital_id: str, case_type: str) -> Optional[dict]:
    if mongodb is None: return None
    
    all_docs = await user_service.get_all_users(role='doctor', hospital_id=hospital_id)
    doctors = [
        item for item in all_docs
        if _doctor_is_available(item) and _doctor_matches_case(item, case_type)
    ]
    if not doctors:
        return None

    # Get active bookings from MongoDB
    cursor = mongodb.advanced_bookings.find({"hospital_id": hospital_id})
    existing = await cursor.to_list(length=1000)

    weighted = []
    for doctor in doctors:
        doctor_id = str(doctor.get('id') or doctor.get('_id') or '')
        active_count = sum(
            1 for item in existing
            if str(item.get('doctor_id') or '') == doctor_id
            and str(item.get('status') or '').lower() in ('allocated', 'in_consultation')
        )
        weighted.append((active_count, str(doctor.get('name') or ''), doctor))

    weighted.sort(key=lambda item: (item[0], item[1]))
    return weighted[0][2]

async def create_booking(data: dict) -> dict:
    if mongodb is None: return {}
    booking_id = _gen_id("AB-")
    now_dt = datetime.utcnow()

    doctor_id = str(data.get("doctor_id") or "")
    doctor_name = data.get("doctor_name", "")
    doctor_specialization = data.get("doctor_specialization", "")

    if not doctor_id:
        allocated_doctor = await _allocate_doctor(str(data.get('hospital_id') or ''), str(data.get('case_type') or ''))
        if not allocated_doctor:
            raise ValueError('No eligible doctor available for this advanced booking.')
        doctor_id = str(allocated_doctor.get('id') or allocated_doctor.get('_id') or '')
        doctor_name = allocated_doctor.get('name', '')
        doctor_specialization = allocated_doctor.get('specialization') or allocated_doctor.get('department', '')

    booking_doc = {
        "_id": booking_id,
        "id": booking_id,
        "case_type": data.get("case_type"),
        "case_label": data.get("case_label"),
        "patient_id": str(data.get("patient_id") or ""),
        "patient_name": data.get("patient_name"),
        "patient_age": data.get("patient_age", 0),
        "patient_gender": data.get("patient_gender", ""),
        "hospital_id": str(data.get("hospital_id") or ""),
        "hospital_name": data.get("hospital_name", ""),
        "doctor_id": doctor_id,
        "doctor_name": doctor_name,
        "doctor_specialization": doctor_specialization,
        "room": data.get("room"),
        "room_id": str(data.get("room_id") or ""),
        "room_type": data.get("room_type", ""),
        "reason": data.get("reason", ""),
        "appointment_date": data.get("appointment_date"),
        "priority": data.get("priority", "high"),
        "status": str(data.get("status", "allocated")),
        "allocation_method": data.get("allocation_method", "auto-doctor-load-balanced"),
        "source": data.get("source", "advanced-booking"),
        "allocated_at": now_dt,
        "created_at": now_dt,
        "updated_at": now_dt,
    }

    await mongodb.advanced_bookings.insert_one(booking_doc)
    return booking_doc

async def get_booking_by_id(booking_id: str) -> Optional[dict]:
    if mongodb is None: return None
    booking = await mongodb.advanced_bookings.find_one({"_id": booking_id})
    if booking:
        booking["id"] = str(booking.get("_id", booking.get("id")))
    return booking

async def get_bookings_scoped(payload: dict) -> List[dict]:
    if mongodb is None: return []
    role = str(payload.get("role") or "").lower()
    user_id = str(payload.get("sub") or "")
    hospital_id = str(payload.get("hospital_id") or "")

    query = {}
    if role == "admin":
        pass
    elif role == "hm":
        query["hospital_id"] = hospital_id
    elif role == "doctor":
        query["doctor_id"] = user_id
    elif role == "patient":
        query["patient_id"] = user_id
    else:
        return []

    cursor = mongodb.advanced_bookings.find(query).sort("created_at", -1)
    bookings = await cursor.to_list(length=1000)
    for b in bookings:
        b["id"] = str(b.get("_id", b.get("id")))
    return bookings

async def update_booking_status(booking_id: str, status: str, payload: Dict[str, Any]) -> Optional[dict]:
    if mongodb is None: return None
    booking = await get_booking_by_id(booking_id)
    if not booking:
        return None

    role = str(payload.get("role") or "").lower()
    actor_id = str(payload.get("sub") or "")
    actor_hospital_id = str(payload.get("hospital_id") or "")

    if role == "doctor" and str(booking.get("doctor_id") or "") != actor_id:
        raise PermissionError("Doctor can update only own allocated bookings")

    if role == "hm" and actor_hospital_id and str(booking.get("hospital_id") or "") != actor_hospital_id:
        raise PermissionError("HM can update only own hospital bookings")

    if role not in ("doctor", "hm", "admin"):
        raise PermissionError("Role not allowed")

    updates = {
        "status": status,
        "updated_at": datetime.utcnow(),
    }

    await mongodb.advanced_bookings.update_one({"_id": booking_id}, {"$set": updates})
    return await get_booking_by_id(booking_id)

async def assign_booking_doctor(
    booking_id: str,
    doctor_id: str,
    payload: Dict[str, Any],
    doctor_name: str = "",
    doctor_specialization: str = "",
    allocation_method: str = "hm-manual-assignment",
) -> Optional[dict]:
    if mongodb is None: return None
    booking = await get_booking_by_id(booking_id)
    if not booking:
        return None

    role = str(payload.get("role") or "").lower()
    actor_hospital_id = str(payload.get("hospital_id") or "")

    if role not in ("hm", "admin"):
        raise PermissionError("Role not allowed")

    if role == "hm" and actor_hospital_id and str(booking.get("hospital_id") or "") != actor_hospital_id:
        raise PermissionError("HM can update only own hospital bookings")

    selected_doctor = await user_service.get_user_by_id(str(doctor_id))
    if not selected_doctor or str(selected_doctor.get("role") or "").lower() != "doctor":
        raise ValueError("Selected doctor not found")

    booking_hospital_id = str(booking.get("hospital_id") or "")
    doctor_hospital_id = str(selected_doctor.get("hospital_id") or "")
    if booking_hospital_id and doctor_hospital_id and booking_hospital_id != doctor_hospital_id:
        raise ValueError("Selected doctor does not belong to this hospital")

    updates = {
        "doctor_id": str(doctor_id),
        "doctor_name": doctor_name or selected_doctor.get("name") or "Doctor",
        "doctor_specialization": doctor_specialization or selected_doctor.get("specialization") or "",
        "allocation_method": allocation_method or "hm-manual-assignment",
        "status": "allocated",
        "updated_at": datetime.utcnow(),
    }

    await mongodb.advanced_bookings.update_one({"_id": booking_id}, {"$set": updates})
    return await get_booking_by_id(booking_id)

File: noq-frontend/src/services/advancedBookingService.js
import api from './api.js';
import firebaseDbService from './firebaseDbService.js';

const COLLECTION_NAME = 'advancedBookings';

const normalizeBooking = (item = {}) => ({
  id: item.id,
  caseType: item.caseType || item.case_type,
  caseLabel: item.caseLabel || item.case_label,
  patientId: String(item.patientId || item.patient_id || ''),
  patientName: item.patientName || item.patient_name,
  patientAge: item.patientAge ?? item.patient_age ?? 0,
  patientGender: item.patientGender || item.patient_gender || '',
  hospitalId: String(item.hospitalId || item.hospital_id || ''),
  hospitalName: item.hospitalName || item.hospital_name || '',
  doctorId: String(item.doctorId || item.doctor_id || ''),
  doctorName: item.doctorName || item.doctor_name || '',
  doctorSpecialization: item.doctorSpecialization || item.doctor_specialization || '',
  room: item.room,
  roomId: String(item.roomId || item.room_id || ''),
  roomType: item.roomType || item.room_type || '',
  reason: item.reason || '',
  appointmentDate: item.appointmentDate || item.appointment_date,
  priority: item.priority || 'high',
  status: item.status || 'allocated',
  allocationMethod: item.allocationMethod || item.allocation_method || '',
  source: item.source || 'advanced-booking',
  allocatedAt: item.allocatedAt || item.allocated_at,
  createdAt: item.createdAt || item.created_at,
  updatedAt: item.updatedAt || item.updated_at,
});

const toApiPayload = (item = {}) => ({
  case_type: item.caseType,
  case_label: item.caseLabel,
  patient_id: String(item.patientId || ''),
  patient_name: item.patientName,
  patient_age: Number(item.patientAge || 0),
  patient_gender: item.patientGender || '',
  hospital_id: String(item.hospitalId || ''),
  hospital_name: item.hospitalName || '',
  doctor_id: String(item.doctorId || ''),
  doctor_name: item.doctorName || '',
  doctor_specialization: item.doctorSpecialization || '',
  room: item.room,
  room_id: String(item.roomId || ''),
  room_type: item.roomType || '',
  reason: item.reason || '',
  appointment_date: item.appointmentDate,
  priority: item.priority || 'high',
  status: item.status || 'allocated',
  allocation_method: item.allocationMethod || '',
  source: item.source || 'advanced-booking',
  allocated_at: item.allocatedAt || new Date().toISOString(),
});

export const advancedBookingService = {
  // Create booking in Firebase (primary source of truth)
  create: async (booking) => {
    const normalized = normalizeBooking(booking);

    try {
      const res = await api.post('/advanced-bookings/create', toApiPayload(normalized));
      if (res?.success) {
        const created = normalizeBooking(res.data);
        await firebaseDbService.upsert(COLLECTION_NAME, created.id, created);
        return created;
      }
    } catch (error) {
      console.warn('API create failed, using Firebase directly:', error);
    }

    // Firebase fallback
    const localItem = {
      ...normalized,
      id: normalized.id || `AB-${Date.now()}`,
      createdAt: normalized.createdAt || new Date().toISOString(),
      updatedAt: normalized.updatedAt || new Date().toISOString(),
    };

    await firebaseDbService.upsert(COLLECTION_NAME, localItem.id, localItem);
    return localItem;
  },

  // Get user's bookings from Firebase
  getMine: async (fallbackFilter = () => true) => {
    try {
      const res = await api.get('/advanced-bookings/mine');
      if (res?.success) {
        const list = (res?.data?.bookings || []).map(normalizeBooking);
        // Sync to Firebase
        for (const item of list) {
          await firebaseDbService.upsert(COLLECTION_NAME, item.id, item);
        }
        return list;
      }
    } catch (error) {
      console.warn('API fetch failed, using Firebase:', error);
    }

    // Firebase fallback
    try {
      const all = await firebaseDbService.getCollection(COLLECTION_NAME);
      return (all || []).filter(fallbackFilter).map(normalizeBooking);
    } catch (error) {
      console.error('Failed to get bookings from Firebase:', error);
      return [];
    }
  },

  // Update booking status in Firebase
  updateStatus: async (bookingId, status) => {
    try {
      const res = await api.patch(`/advanced-bookings/${bookingId}/status`, { status });
      if (res?.success) {
        const updated = normalizeBooking(res.data);
        await firebaseDbService.upsert(COLLECTION_NAME, bookingId, updated);
        return updated;
      }
    } catch (error) {
      console.warn('API update failed, using Firebase:', error);
    }

    // Firebase fallback
    const updated = {
      status,
      updatedAt: new Date().toISOString(),
    };
    await firebaseDbService.upsert(COLLECTION_NAME, bookingId, updated);
    return await firebaseDbService.getDocument(COLLECTION_NAME, bookingId);
  },
};

export default advancedBookingService;
